osgtest/tests/test_460_stashcache.py

```diff
@@ -0,0 +1,94 @@
+import os
+import pwd
+import random
+import tempfile
+
+from osgtest.library import core
+from osgtest.library import files
+from osgtest.library.osgunittest import OSGTestCase
+from osgtest.library import service
+try:
+    from urllib2 import urlopen
+except ImportError:
+    from urllib.request import urlopen
+
+
+_NAMESPACE = "stashcache"
+
+def _getcfg(key):
+    return core.config["%s.%s" % (_NAMESPACE, key)]
+
+def _setcfg(key, val):
+    core.config["%s.%s" % (_NAMESPACE, key)] = val
+
+
+class TestStashCache(OSGTestCase):
+    # testfiles with random contents
+    testfiles = [
+        ("testfile%d" % x, str(random.random()))
+        for x in range(4)
+    ]
+
+    def assertCached(self, name, contents):
+        fpath = os.path.join(_getcfg("cache_dir"), name)
+        self.assertTrue(os.path.exists(fpath),
+                        name + " not cached")
+        self.assertEqual(contents, files.read(fpath, as_single_string=True),
+                         "cached file %s contents do not match expected" % name)
+
+    def skip_bad_unless_running(self, *services):
+        for svc in services:
+            self.skip_bad_unless(service.is_running(svc), "%s not running" % svc)
+
+    @core.elrelease(7,8)
+    def setUp(self):
+        core.skip_ok_unless_installed("stashcache-origin-server", "stashcache-cache-server", "stashcache-client")
+        self.skip_bad_unless_running("xrootd@stashcache-origin-server", "xrootd@stashcache-cache-server")
+
+    def test_01_create_files(self):
+        xrootd_user = pwd.getpwnam("xrootd")
+        for name, contents in self.testfiles:
+            files.write(os.path.join(_getcfg("origin_dir"), name),
+                        contents, backup=False, chmod=0o644,
+                        chown=(xrootd_user.pw_uid, xrootd_user.pw_gid))
+
+    def test_02_xroot_fetch_from_origin(self):
+        name, contents = self.testfiles[0]
+        result, _, _ = \
+            core.check_system(["xrdcp", "-d1", "-N", "-f",
+                               "root://localhost:%d//%s" % (_getcfg("origin_xroot_port"), name),
+                               "-"], "Checking xroot copy from origin")
+        self.assertEqual(result, contents, "downloaded file does not match expected")
```

Comment:
Can we test HTTP fetches from the origin or is that not supported?

Reply:
The origin isn't running the HTTP plugin.